staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1019 @@
+			ca.setRemoteIfc(newIfc, ifcKey)
+		}
+	}
+	return err
+}
+
+func (ca *ConnectionAgent) setHostIPAndIfcNameInAttStatus(att *netv1a1.NetworkAttachment, ifcName string) (*netv1a1.NetworkAttachment, error) {
+	att2 := att.DeepCopy()
+	att2.Status.HostIP = ca.hostIP
+	att2.Status.IfcName = ifcName
+	att3, err := ca.netv1a1Ifc.NetworkAttachments(att2.Namespace).Update(att2)
+	return att3, err
+}
+
+func (ca *ConnectionAgent) ifcUpdateIsNeeded(ifc netfabric.NetworkInterface, att *netv1a1.NetworkAttachment) bool {
+	if ifc.GuestIP.String() != att.Status.IPv4 {
+		return true
+	}
+	if ifc.HostIP.String() != att.Status.HostIP {
+		return true
+	}
+	if ifc.VNI != att.Spec.VNI {
+		return true
+	}
+	return false
+}
+
+func (ca *ConnectionAgent) deleteLocalIfc(ifcKey vniAndNsn) (ifc netfabric.NetworkInterface, ifcFound bool, err error) {
+	ifc, ifcFound = ca.getLocalIfc(ifcKey)
+	if !ifcFound {
+		return
+	}
+	err = ca.netFabric.DeleteLocalIfc(ifc)
+	if err == nil {
+		ca.unsetLocalIfc(ifcKey)
+	}
+	return
+}
+
+func (ca *ConnectionAgent) deleteRemoteIfc(ifcKey vniAndNsn) (ifc netfabric.NetworkInterface, ifcFound bool, err error) {
+	ifc, ifcFound = ca.getRemoteIfc(ifcKey)
+	if !ifcFound {
+		return
+	}
+	err = ca.netFabric.DeleteRemoteIfc(ifc)
+	if err == nil {
+		ca.unsetRemoteIfc(ifcKey)
+	}
+	return
+}
+
+func (ca *ConnectionAgent) setLocalIfc(ifc netfabric.NetworkInterface, ifcKey vniAndNsn) {
+	ca.localIfcsMutex.Lock()
+	defer ca.localIfcsMutex.Unlock()
+	ca.localIfcs[ifcKey] = ifc
+	return
+}
+
+func (ca *ConnectionAgent) setRemoteIfc(ifc netfabric.NetworkInterface, ifcKey vniAndNsn) {
+	ca.remoteIfcsMutex.Lock()
+	defer ca.remoteIfcsMutex.Unlock()
+	ca.remoteIfcs[ifcKey] = ifc
+	return
+}
+
+func (ca *ConnectionAgent) unsetLocalIfc(ifcKey vniAndNsn) {
+	ca.localIfcsMutex.Lock()
+	defer ca.localIfcsMutex.Unlock()
+	delete(ca.localIfcs, ifcKey)
+}
+
+func (ca *ConnectionAgent) unsetRemoteIfc(ifcKey vniAndNsn) {
+	ca.remoteIfcsMutex.Lock()
+	defer ca.remoteIfcsMutex.Unlock()
+	delete(ca.remoteIfcs, ifcKey)
+}
+
+func (ca *ConnectionAgent) getLocalIfc(ifcKey vniAndNsn) (ifc netfabric.NetworkInterface, ifcFound bool) {
+	ca.localIfcsMutex.RLock()
+	defer ca.localIfcsMutex.RUnlock()
+	ifc, ifcFound = ca.localIfcs[ifcKey]
+	return
+}
+
+func (ca *ConnectionAgent) getRemoteIfc(ifcKey vniAndNsn) (ifc netfabric.NetworkInterface, ifcFound bool) {
+	ca.remoteIfcsMutex.RLock()
+	defer ca.remoteIfcsMutex.RUnlock()
+	ifc, ifcFound = ca.remoteIfcs[ifcKey]
+	return
+}
+
+func (ca *ConnectionAgent) fromAttToEnquableRef(att *netv1a1.NetworkAttachment) (enquableRef attQueueRef) {
+	enquableRef.nsn = attNSN(att)
+	enquableRef.vni = att.Spec.VNI
+	if att.Spec.Node == ca.localNodeName {
+		enquableRef.local = true
+	} else {
+		enquableRef.local = false
+	}
+	return
+}
+
+// Return a string representing a field selector that matches NetworkAttachments
+// that run on the local node and have a virtual IP.
+func (ca *ConnectionAgent) localAttWithAnIPSelector() string {
```

Comment:
Rather than working directly with the syntax, I think it would be better to use https://github.com/kubernetes/apimachinery/blob/master/pkg/fields/selector.go to construct a data structure and conver it to one string at the end.

Reply:
I agree but unfortunately field selectors pkg is not expressive enough in release 1.10. We can implement what you're proposing when we move to release 1.11 or higher.